### PR TITLE
[kmac] Add Error checking logic

### PR DESCRIPTION
**Pull request by eunchan**
This commit combines scattered error checking logics into a single
module. Still a few errors are checked in the submodules and reported
from there.

This is the follow-up commit for issue #6370 #6635 #6638

**Pending Items**
- [x] Revise SHA3 error code consistent (double process may / may not trigger error in current senario)
- [ ] ~~Remove SHA3 error report for KMAC (And waive from coverage too)~~
- [x] Revise KMAC Error handling section in the spec

kmac_err
kmac_err_blockdiagram

### CODE REVIEW

File: hw/ip/kmac/rtl/kmac.sv [hunk 1]
```diff
@@ -802,6 +812,31 @@ module kmac
...
+  kmac_errchk u_errchk (
+    .clk_i,
+    .rst_ni,
+
+    // Configurations
+    .cfg_mode_i    (reg_sha3_mode      ),
+    .cfg_strength_i(reg_keccak_strength),
+
+    .kmac_en_i      (reg_kmac_en        ),
+    .cfg_prefix_6B_i(reg_ns_prefix[47:0]), // first 6B of PREFIX
```
Comment (udinator) [inline hw/ip/kmac/rtl/kmac.sv:825]: nit : maybe name this `fname_i` or something like that, since this should only be the function name bits?

File: hw/ip/kmac/rtl/kmac_errchk.sv [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+    end
+  end : check_modestrength
+
+
+  // Check prefix 6B is `encode_string("KMAC")`
+  always_comb begin : check_prefix
+    err_prefix = 1'b 0;
+
+    if (st == StIdle && st_d == StMsgFeed && kmac_en_i) begin
+      if (cfg_prefix_6B_i != 48'h 4341_4D4B2001) begin
```
Comment (udinator) [inline hw/ip/kmac/rtl/kmac_errchk.sv:182]: should this value be a parameter?

Comment (udinator) [reply hw/ip/kmac/rtl/kmac_errchk.sv:182]: nit : `48'h 4341_4D4B2001` -> `48'h 4341_4D4B_2001`

File: hw/ip/kmac/rtl/kmac_errchk.sv [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+  /////////////
+  // Signals //
+  /////////////
+
+  // `err_swsequence` occurs when SW issues wrong command
+  logic err_swsequence;
+
+  // `err_modestrength` occcurs when Mode & Strength combinations are not
+  // allowed. This error does not block the hashing operation.
+  logic err_modestrength;
```
Comment (udinator) [inline hw/ip/kmac/rtl/kmac_errchk.sv:105]: if incorrectly configuring KMAC does not reset everything back to Idle state, what is the expected behavior?
will the IP keep these incorrect configs and still hash the msg?

File: hw/ip/kmac/rtl/kmac_errchk.sv [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+
+  // `err_prefix` occurs when the first 6B of !!PREFIX is not
+  // `encode_string("KMAC")` and kmac is enabled. This error does not block the
+  // KMAC operation.
+  logic err_prefix;
```
Comment (udinator) [inline hw/ip/kmac/rtl/kmac_errchk.sv:110]: similar question here (though probably less severe than the above) - will the IP keep hashing using the incorrect function name?

File: hw/ip/kmac/rtl/kmac.sv (hunk 1 above)
Comment (eunchan) [reply hw/ip/kmac/rtl/kmac.sv:825]: the prefix may have bigger function name bytes. let me think. if fname is appropriate.

File: hw/ip/kmac/rtl/kmac_errchk.sv (hunk 3 above)
Comment (eunchan) [reply hw/ip/kmac/rtl/kmac_errchk.sv:105]: it will get incorrect digest. SW will know it used wrong mode & strength. The reason why I let it move is to give an opportunity for SW to run other combinations (maybe later to use cSHAKE512?)

File: hw/ip/kmac/rtl/kmac_errchk.sv (hunk 4 above)
Comment (eunchan) [reply hw/ip/kmac/rtl/kmac_errchk.sv:110]: yes.

File: hw/ip/kmac/rtl/kmac_errchk.sv (hunk 2 above)
Comment (eunchan) [reply hw/ip/kmac/rtl/kmac_errchk.sv:182]: SG, Let me define `encoded_string("KMAC")`

File: hw/ip/kmac/rtl/kmac.sv (hunk 1 above)
Comment (udinator) [reply hw/ip/kmac/rtl/kmac.sv:825]: hmm, actually good point - what if someone writes `"KMACHMAC"` as the function string, then this will bypass the error?

**Comment by eunchan** [comment]
Document has been revised too (Added Error report section)

File: hw/ip/kmac/rtl/kmac_app.sv [hunk 5]
```diff
@@ -487,22 +490,20 @@ module kmac_app
...
-          code: ErrSwPushedWrongCmd,
-          info: 24'(sw_cmd_i)
-        };
-      end
+    if (mux_sel != SelSw && sw_valid_i) begin
+      // If SW writes message into FIFO
+      mux_err = '{
+        valid: 1'b 1,
+        code: ErrSwPushedMsgFifo,
+        info: 24'({8'h 00, 8'(st), 8'(mux_sel)})
```
Comment (udinator) [inline hw/ip/kmac/rtl/kmac_app.sv:498]: can the enums for the states and mux_sel be placed in `kmac_pkg` ?
since these are used in the error reporting, the scoreboard will need access to these values which is more convenient for DV if they are in the common IP package.

**Comment by eunchan** [comment]
Thanks for the suggestion. Moved to `kmac_pkg::app_mux_sel_e`.

**Comment by udinator** [comment]
> Thanks for the suggestion. Moved to `kmac_pkg::app_mux_sel_e`.

Thanks! Could you also move the FSM state enum as well?

File: hw/ip/kmac/doc/_index.md [hunk 6]
```diff
@@ -256,6 +256,12 @@ The refresh does not block the internal entropy expansion operation.
 
 ### Error Report
 
+This section explains the errors KMAC HWIP raises during the hasing operations, their meanings, and the error handling process.
+
+KMAC HWIP has the error checkers in its internal datapath.
+If the checkers detect errors, whether they are triggered by the SW mis-configure, or HW malfunctions, they report the error to !!ERR_CODE and raises an `kmac_error` interrupt.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md]: nit: raises -> raise

File: hw/ip/kmac/rtl/kmac_errchk.sv [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+  st_e st, st_d;
+
+  /////////////
+  // Signals //
+  /////////////
+
+  // `err_swsequence` occurs when SW issues wrong command
+  logic err_swsequence;
+
+  // `err_modestrength` occcurs when Mode & Strength combinations are not
```
Comment (msfschaffner) [inline hw/ip/kmac/rtl/kmac_errchk.sv:103]: nit: occcurs -> occurs

File: hw/ip/kmac/rtl/kmac_errchk.sv [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+
+
+  ///////////////////
+  // State Machine //
+  ///////////////////
+  always_ff @(posedge clk_i or negedge rst_ni) begin
+    if (!rst_ni) begin
+      st <= StIdle;
+    end else begin
+      st <= st_d;
```
Comment (msfschaffner) [inline hw/ip/kmac/rtl/kmac_errchk.sv:251]: nit: shouldn't this be `st_q <= st_d`?

File: hw/ip/kmac/doc/_index.md [hunk 9]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+This section explains the errors KMAC HWIP raises during the hasing operations, their meanings, and the error handling process.
+
+KMAC HWIP has the error checkers in its internal datapath.
+If the checkers detect errors, whether they are triggered by the SW mis-configure, or HW malfunctions, they report the error to {{< regref "ERR_CODE" >}} and raises an `kmac_error` interrupt.
+Each error code gives debugging information at the lower 24 bits of {{< regref "ERR_CODE" >}}.
+
+Value | Error Code | Description
+------|------------|-------------
+0x01  | KeyNotValid | In KMAC mode with the sideloaded key, the IP raises an error if the sideloaded secret key is not ready.
+0x02  | SwPushedMsgFifo | MsgFifo is updated not in the Message Feed state.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:268]: Maybe slightly rephrase to `MsgFifo is updated while not being in the Message Feed state.`?

File: hw/ip/kmac/doc/_index.md [hunk 10]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+0x03  | SwIssuedCmdInAppActive | SW issued a command while the application interface is being used
+0x04  | WaitTimerExpired | EDN has not responded within the wait timer limit.
+0x05  | IncorrectEntropyMode | When SW sets `entropy_ready`, the `entropy_mode` is neither SW nor EDN.
+0x06  | UnexpectedModeStrength | SHA3 mode and Keccak Strength combination is not expected.
+0x07  | IncorrectFunctionName | In KMAC mode, the PREFIX has the value other than `encoded_string("KMAC")`
+0x08  | SwCmdSequence | Sw does not follow the guided sequence, `start` -> `process` -> {`run` ->} `done`
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:274]: Nit: Sw -> SW

File: hw/ip/kmac/doc/_index.md [hunk 11]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+0x80  | Sha3Control | SW may receive Sha3Control error along with `SwCmdSequence` error. Can be ignored.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:275]: Aren't these codes mutually exclusive due to priority encoding?

File: hw/ip/kmac/doc/_index.md [hunk 12]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+#### KeyNotValid (0x01)
+
+The `KeyNotValid` error is raised in the application interface module.
+When a KMAC application requests a hashing operation, the module checks if the sideloaded key is ready.
+If the key is not ready, the module reports `KeyNotValid` error and moves to dead-end state and waits the IP reset.
+
+This eror does not provide any additional information.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:283]: nit: eror -> error

File: hw/ip/kmac/doc/_index.md [hunk 13]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+#### SwPushedMsgFifo (0x02)
+
+The `SwPushedMsgFifo` error happens when the Message FIFO recevies TL-UL transactions while the application interface is in active.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:287]: the last parts of this sentence may be misleading as it could be interpreted as "inactive" (and I believe this would be wrong, right?).

how about "while the application interface is busy" or "in an active state"?

File: hw/ip/kmac/doc/_index.md [hunk 14]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+Bits    | Name        | Description
+--------|-------------|-------------
+[23:16] | reserved    | all zero
+[15:8]  | kmac_app_st | KMAC_APP FSM state.
+[7:0]   | mux_sel     | Current APP Mux selection. 0: None, 1: SW, 2: App
+
+#### SwIssuedCmdInAppActive (0x03)
+
+If the SW issues any commands during the application interface is being used, the module reports `SwIssuedCmdInAppActive` error.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:300]: during -> while

File: hw/ip/kmac/doc/_index.md [hunk 15]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+The request is dropped by the KMAC_APP module.
+
+The lower 3 bits of {{< regref "ERR_CODE" >}} contains the received command from the SW.
+#### WaitTimerExpired (0x04)
+
+The SW may set the EDN wait timer to exit from EDN request state if the response from EDN takes long.
+If the timer expires, the module cancels the transaction and report the `WaitTimerExpired` error.
+
+When this error happens, the state machine in KMAC_ENTROPY module moves to Wait state.
+In the state, it keeps using the pre-generated entropy and asserting the entropy valid signal.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:311]: nit: I would probably say "in *that* state"

File: hw/ip/kmac/doc/_index.md [hunk 16]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+#### IncorrectEntropyMode (0x05)
+
+If SW misconfigures the entropy mode and let the entropy module prepare the random data, the module reports `IncorrectEntropyMode` error.
+The state machine moves to Wait state after reporting the error.
+
+The SW may move the state machine to the reset state by issuing {{<regref "CFG.err_processed" >}}.
+
+#### UnexpectedModeStrength (0x06)
+
+When the SW issues `Start` command, the KMAC_ERRCHK module checks the {{< regref "CFG.mode" >}} and {{< regref "CFG.kstrength" >}}.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:327]: nit: I would probably reformulate as 

When the SW issues `Start` command, the KMAC_ERRCHK module checks the {{< regref "CFG.mode" >}} and {{< regref "CFG.kstrength" >}} values and reports a `UnexpectedModeStrength` error if the combination of these values is unexpected. Expected combinations are listed below:
- **SHA3-224**
- ....

Comment (msfschaffner) [reply hw/ip/kmac/doc/_index.md:327]: EDIT: not sure but we could also call this error `NonStandardModeStrength` if this list refers to standardized combinations.

File: hw/ip/kmac/doc/_index.md [hunk 17]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+The KMAC HWIP assumes the combinations of two to be **SHA3-224**, **SHA3-256**, **SHA3-384**, **SHA3-512**, **SHAKE-128**, **SHAKE-256**, **cSHAKE-128**, and **cSHAKE-256**.
+If the combination of the `mode` and `kstrength` does not fall into above, the module reports the `UnexpectedModeStrength` error.
+
+However, the KMAC HWIP proceeds the hashing operation as other combinations does not cause any malfunctions inside the IP.
+The SW may get the incorrect digest value.
+
+#### IncorrectFunctionName (0x07)
+
+If {{< regref "CFG.kmac_en" >}} is set and the SW issues the `Start` command, the KMAC_ERRCHK checks if the {{< regref "PREFIX" >}} has correct function name, `encode_string("KMAC")`.
+If the value does not match to the byte form of `encode_string("KMAC")` (0x4341_4D4B_2001), it reports the `IncorrectFunctionName` error.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:337]: not sure, but should we `tick` the hex value here?

File: hw/ip/kmac/doc/_index.md [hunk 18]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+As same as `UnexpectedModeStrength` error, this error does not block the hashing operation.
+The SW may get the incorrect signature value.
+
+#### SwCmdSequence (0x08)
+
+The KMAC_ERRCHK module checks the SW issued commands if it follows the guideline.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:344]: nit: I would rephrase to something like
"The KMAC_ERRCHK module checks whether the SW issued commands follows the recommended programming sequence."

File: hw/ip/kmac/doc/_index.md [hunk 19]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+If the SW issues the command that is not relavant to the current context, the module reports the `SwCmdSequence` error.
+The lower 3bits of the {{< regref "ERR_CODE" >}} contains the received command.
+
+This error, however, may not stop the KMAC HWIP.
```
Comment (msfschaffner) [inline hw/ip/kmac/doc/_index.md:348]: "may" or "will"?

**Review by msfschaffner — COMMENTED** [review]
Thanks @eunchan - my comments are mostly editorial.

File: hw/ip/kmac/doc/_index.md [hunk 20]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+The `KeyNotValid` error is raised in the application interface module.
+When a KMAC application requests a hashing operation, the module checks if the sideloaded key is ready.
+If the key is not ready, the module reports `KeyNotValid` error and moves to dead-end state and waits the IP reset.
+
+This error does not provide any additional information.
+
+#### SwPushedMsgFifo (0x02)
+
+The `SwPushedMsgFifo` error happens when the Message FIFO recevies TL-UL transactions while the application interface is busy.
```
Comment (cdgori) [inline hw/ip/kmac/doc/_index.md:287]: nit: receives

File: hw/ip/kmac/doc/_index.md [hunk 21]
```diff
@@ -256,7 +256,97 @@ The refresh does not block the internal entropy expansion operation.
...
+
+The lower 3 bits of {{< regref "ERR_CODE" >}} contains the received command from the SW.
+#### WaitTimerExpired (0x04)
+
+The SW may set the EDN wait timer to exit from EDN request state if the response from EDN takes long.
+If the timer expires, the module cancels the transaction and report the `WaitTimerExpired` error.
+
+When this error happens, the state machine in KMAC_ENTROPY module moves to Wait state.
+In that state, it keeps using the pre-generated entropy and asserting the entropy valid signal.
+It asserts the entropy valid signal to complete the current hasing operation.
```
Comment (cdgori) [inline hw/ip/kmac/doc/_index.md:312]: nit: hashing

File: hw/ip/kmac/rtl/kmac.sv [hunk 22]
```diff
@@ -437,21 +442,38 @@ module kmac
...
-      hw2reg.err_code.d = {keymgr_err.code, keymgr_err.info};
-    end else if (entropy_err.valid) begin
-      hw2reg.err_code.d = {entropy_err.code, entropy_err.info};
-    end else begin
-      hw2reg.err_code.d = '0;
-    end
+    hw2reg.err_code.d = '0;
+
+    priority case (1'b 1)
+      // app_err has the highest priority. If SW issues wrong command while app
```
Comment (cdgori) [inline hw/ip/kmac/rtl/kmac.sv:455]: maybe rephrase this comment (not sure if what I've written is correct, but I found the original version ambiguous):

// app_err has the highest priority.  If software issues an incorrect command while the app
// is in the active state, the error from AppIntf is passed through.

**Comment by eunchan** [comment]
Another CI hiccups here @imphil . This  PR was passed and just rebased to the upstream/master and CI failed with 

```
 ---> Running in 452ee03c4848
/bin/sh: 1: pip3: not found
```


message in "Build Docker Image" step.

**Comment by rswarbrick** [comment]
The runner appears not to have been able to connect to ubuntu.com. Probably best to fix the private CI failure and then force-push. Hopefully, "IT weather" will be clearer by then. If not, we've got some work to do tomorrow.

File: hw/ip/kmac/dv/env/kmac_env_cov.sv [hunk 23]
```diff
@@ -308,7 +308,7 @@ class kmac_env_cov extends cip_base_env_cov #(.CFG_T(kmac_env_cfg));
       bins err_none = {kmac_pkg::ErrNone};
       bins err_key_not_valid = {kmac_pkg::ErrKeyNotValid};
       bins err_sw_pushed_msg_fifo = {kmac_pkg::ErrSwPushedMsgFifo};
-      bins err_sw_pushed_wrong_cmd = {kmac_pkg::ErrSwPushedWrongCmd};
+      bins err_sw_pushed_wrong_cmd = {kmac_pkg::ErrSwIssuedCmdInAppActive};
```
Comment (udinator) [inline hw/ip/kmac/dv/env/kmac_env_cov.sv:311]: nit: can you change the bins name to `err_sw_issued_cmd_in_app_active` to match?

**Comment by eunchan** [comment]
I think I have addressed the comments. Can anyone approve this? :)

**Review by udinator — APPROVED** [review]
LGTM overall, lets get this RTL update merged.
I will check later today/tmrw whether the issues I linked in #6676 are still present.

**Comment by eunchan** [comment]
Hum.. again CI failure after rebasing... :(

**Comment by eunchan** [comment]
@imphil @rswarbrick could you please check if NexysVideo FPGA on CI runs OK? I didn't change SPI_D, HMAC, UART at all in this PR but somehow SPI frame cannot be transferred correctly.

**Comment by eunchan** [comment]
OK it works now and passed the CI. Just merged
